processing/src/main/java/io/druid/segment/filter/InFilter.java

```diff
@@ -23,30 +23,45 @@
 import com.google.common.base.Predicate;
 import com.google.common.base.Strings;
 import com.google.common.collect.Iterables;
+import com.google.common.primitives.Longs;
 import com.metamx.collections.bitmap.ImmutableBitmap;
 import io.druid.query.extraction.ExtractionFn;
 import io.druid.query.filter.BitmapIndexSelector;
+import io.druid.query.filter.DruidLongPredicate;
+import io.druid.query.filter.DruidPredicateFactory;
 import io.druid.query.filter.Filter;
-import io.druid.query.filter.RowOffsetMatcherFactory;
 import io.druid.query.filter.ValueMatcher;
 import io.druid.query.filter.ValueMatcherFactory;
-import io.druid.segment.column.ColumnCapabilities;
 
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
 import java.util.Set;
 
 /**
  */
 public class InFilter implements Filter
 {
+  // determined through benchmark that binary search on long[] is faster than HashSet until ~16 elements
+  // Hashing threshold is not applied to String for now, String still uses ImmutableSortedSet
+  public static final int LONG_HASHING_THRESHOLD = 16;
+
   private final String dimension;
   private final Set<String> values;
   private final ExtractionFn extractionFn;
 
+  private boolean useLongHash;
+  private long[] longArray;
+  private HashSet<Long> longHashSet;
+
   public InFilter(String dimension, Set<String> values, ExtractionFn extractionFn)
   {
     this.dimension = dimension;
     this.values = values;
     this.extractionFn = extractionFn;
+    setLongValues();
```

Comment:
Hmm, it might be worth only doing this if a long predicate is actually requested. I could see the parsing taking a while for long IN filters. But make sure we only do this once even if `getLongPredicate` is called many times.